### ISSUE TITLE
waitTimeout does not work

### ISSUE THREAD

**Issue by cevek** [segmentio/nightmare#375]

**Comment by sporkmonger** [comment]
:+1: I'm also seeing this.

**Comment by matthewmueller** [comment]
can you try again? i just released a new version

**Comment by cevek** [comment]
Cool :+1:

**Comment by contra** [comment]
Also seeing this :rotating_light:

**Comment by matthewmueller** [comment]
@contra ahh yah, you're seeing it hang on the initial page load, right? right now it's only initialized when you call `.wait()`. that should be fixed.

**Comment by YueShuXie** [comment]
Also seeing this.

**Comment by joychester** [comment]
also see this hang forever for own my negative tests(On Mac OS Captian EI + 2.3.2 release), when wait a *non-exist element on page* or wait a *function which does not return true condition* (BTW Page load has already done), 
Please let me know if need my help on reproducing this issue.

**Comment by rosshinkley** [comment]
@contra Depending on the kind of page load failure, #553 _might_ address your issue.

@joychester I wrote two quick test cases:

**nonexistent element**
```js
var Nightmare = require('nightmare'),
  nightmare = Nightmare();

  nightmare.goto('http://www.example.com')
    .wait('does-not-exist')
    .end()
    .then(function(){
      console.log('done');
    })
    .catch(function(err){
      console.dir(err);
    });
```

**function that never returns true**
```js
var Nightmare = require('nightmare'),
  nightmare = Nightmare();

  nightmare.goto('http://www.example.com')
    .wait(function(){return false;})
    .end()
    .then(function(){
      console.log('done');
    })
    .catch(function(err){
      console.dir(err);
    });
```

Both error out with the appropriate error message after 30s.  Can you provide a minimal example that causes nightmare to hang?

**Comment by joychester** [comment]
Hi, @rosshinkley , your case works fine from my side too, but my test pages is not kind of different with yours :)
my page will load the page content async(like SPA application), so if i write wait method like following, it will hang forever:
```javascript
var Nightmare = require('nightmare'),
  nightmare = Nightmare({waitTimeout: 10000, show: true, openDevTools: false});

  nightmare.goto('https://www.stubhub.com/')
    .wait('.event-title123') //this element does not exist..
    .end()
    .then(function(){
      console.log('done');
    })
    .catch(function(err){
      console.dir(err);
    });
```

however, if i hard-wait for the page loaded complete, then the wait() works fine until it timedout(10 seconds in my case)
```javascript
var Nightmare = require('nightmare'),
  nightmare = Nightmare({waitTimeout: 10000, show: true, openDevTools: false});

  nightmare.goto('https://www.stubhub.com/')
    .wait(5000) // wait for enough time until page load completely
    .wait('.event-title123') //this element does not exist..
    .end()
    .then(function(){
      console.log('done');
    })
    .catch(function(err){
      console.dir(err);
    });
```
Seems, nightmare API does not support SPA properly(wait for a element created after Dom content loaded event). Hope this case helps for your further analysis. just let me know if you need any info.

Thanks!

**Comment by rosshinkley** [comment]
@joychester Ah!  Thank you for providing an example.  I can reproduce the behavior you're seeing.  I'll dig into it and see what is going on.

**Comment by rosshinkley** [comment]
@joychester A quick look yields that `webBrowser.evaluateJavaScript` is hanging once in a while and not calling back.  Electron doesn't have any errors when it happens.  This problem is not unique to StubHub - it happens with Yahoo and Youtube as well.  I'm not 100% sold that waiting for an element created after DOM content loaded is the cause, or that the site is an SPA.  

(As an aside, I can hack around it by forcing the `javascript` handler to poll if the callback has been called, but that is terrifyingly hacky.) 

Some tinkering shows me that it's not necessarily iframes, slow loads, load failures, or particularly large assets (or any combination).  I'm still hunting for a root cause/minimum reproduction case.

**Comment by joychester** [comment]
@rosshinkley , Thanks for your info , is that possible to be an issue with Electron? hack way may not elegant, but once you have more insights, please let us know :)

**Comment by rosshinkley** [comment]
@Mr0grog set me straight: it does have to do with iframes and timing.  Quote:

> If you call WebContents’s executeJavaScript() method after the main frame finishes loading, but while a subframe is loading, the JS will never get executed and the call will never complete (unless you navigate the main frame to some new page).

See electon/electron#5183 for details.

**Comment by joychester** [comment]
Thanks @rosshinkley for all your effort and investigation! then we need to wait for Electron permanent fixes.

**Comment by joychester** [comment]
just saw electron/electron#5183 has been closed, which is awesome, looking forward to integrate with Nightmare release and kick off a regression test!! :)

**Comment by Mr0grog** [comment]
Well, Electron has to ship, first ;)  Hopefully soon!

**Comment by joychester** [comment]
they are pretty good team to get things resolved fast, and thank you @Mr0grog and @rosshinkley for chasing them :)

**Comment by joychester** [comment]
Good news, i saw the new electron-0.37.7 has been released, with this electron/electron#5183 fix, change log: https://github.com/electron/electron/releases/tag/v0.37.7

hope we can release nightmare soon with  electron-0.37.7 build 👍

**Comment by rosshinkley** [comment]
Should be fixed with #588.  Closing.

**Comment by joychester** [comment]
verified, 👍
